ts/pages/mesh/liquidity_market.tsx

```diff
@@ -0,0 +1,419 @@
+            return price;
+        },
+        getSize: (order: SignedOrder, op: Operation) => {
+            let size;
+            const takerAssetAmount = new BigNumber(order.takerAssetAmount);
+            const makerAssetAmount = new BigNumber(order.makerAssetAmount);
+            const constant = new BigNumber('1.0e18');
+
+            // tslint:disable-next-line:switch-default
+            switch (op) {
+                case 'ask':
+                    size = takerAssetAmount.div(constant);
+                    break;
+                case 'bid':
+                    size = makerAssetAmount.div(constant);
+            }
+
+            return size;
+        },
+    },
+];
+
+export const LiquidityMarket: React.FC = () => {
+    const [selectedMarketIdx, setSelectedMarketIdx] = useState(0);
+    const [bids, setBids] = useState<OrderData[]>([]);
+    const [asks, setAsks] = useState<OrderData[]>([]);
+    const [spread, setSpread] = useState<BigNumber>();
+
+    const resetData = () => {
+        setBids([]);
+        setAsks([]);
+        setSpread(undefined);
+    };
+
+    useEffect(() => {
+        const fetchData = async () => {
+            const market = markets[selectedMarketIdx];
+            const orders = await fetchOrders(market.baseAsset, market.quoteAsset);
+
+            const _bids = parseOrders(orders.bids.records, 'bid');
+            const _asks = parseOrders(orders.asks.records, 'ask');
+
+            const midPrice = calcMidPrice(_bids, _asks);
+
+            calcSlippage(_bids, midPrice);
+            calcSlippage(_asks, midPrice);
+
+            const _spread = calcSpread(_bids, _asks);
+
+            // Reverse asks for display (price descending order)
+            _asks.reverse();
+
+            setBids(_bids);
+            setAsks(_asks);
+            setSpread(_spread);
+        };
+
+        resetData();
+
+        // tslint:disable-next-line:no-floating-promises
+        fetchData();
+    }, [selectedMarketIdx]);
+
+    const parseOrders = (records: OrderbookRecord[], op: Operation) => {
+        const { getPrice, getSize, orderSizeThreshold = 0 } = markets[selectedMarketIdx];
+
+        const mappedRecords = records.map(({ order }) => ({
+            order,
+            price: getPrice(order, op),
+            size: getSize(order, op),
+        }));
+
+        const filteredRecords = mappedRecords.filter(order => order.size.gte(orderSizeThreshold));
+
+        // if there are more filtered records than required, return them
+        if (filteredRecords.length >= ORDER_COUNT) {
+            return filteredRecords.slice(0, ORDER_COUNT);
+        }
+
+        // else fall back to all retrieved records and return first
+        return mappedRecords.slice(0, ORDER_COUNT);
+    };
+
+    const calcMidPrice = (_bids: OrderData[], _asks: OrderData[]) => {
+        const highestBidPrice = _bids[0].price;
+        const lowestAskPrice = _asks[0].price;
+
+        return highestBidPrice.plus(lowestAskPrice).dividedBy(2);
```

Comment:
changed function in this way to account for one-sided markets (there's certainly a way to make it more elegant)

```
const calcMidPrice = (_bids: OrderData[], _asks: OrderData[]) => {
        let midPrice;
        if (_bids.length === 0) {
            if (_asks.length > 0) {
                midPrice = _asks[0].price;
            }
        }
        if (_asks.length === 0) {
            if (_bids.length > 0) {
                midPrice = _bids[0].price;
            }
        }
        if (midPrice === undefined) {
            midPrice = (_bids[0].price.plus(_asks[0].price).dividedBy(2));
        }

        return midPrice;
    };
```

Also to change calcSpread (next comment)